### ISSUE TITLE
Copying of `.git` into Dockerfile

### ISSUE THREAD

**Issue by candleindark** [datalad/datalad-registry#335]
https://github.com/datalad/datalad-registry/blob/8a998bb0cc39bbbf10134c080b9f8f52b3f6f445/Dockerfile#L54

As illustrated above, currently, the `.git` directory is copied to the docker image of the project because `versioningit`, the versioning system, requires it to build the project. Look in to whether this can be circumvented or avoided.

This issue is related to https://github.com/datalad/datalad-registry/issues/334 to some degree.

**Comment by yarikoptic** [comment]
while doing full "source distribution" all needed version placeholders are provided somewhere, thus avoiding need for carrying `.git` around, right @jwodder ?

**Comment by jwodder** [comment]
@yarikoptic If you're installing from an sdist (assuming that's what you're referring to), then `versioningit` doesn't need (or use) `.git/`.

**Comment by yarikoptic** [comment]
yes, that is what I mean.   so instead of all the manual individual COPY steps there should just be an `sdist` step in https://github.com/datalad/datalad-registry/blob/8a998bb0cc39bbbf10134c080b9f8f52b3f6f445/Dockerfile#L49 with `pip` following installation from it, to make it all clean, right?

**Comment by jwodder** [comment]
@yarikoptic

* If you mean a step that copies an sdist into the Dockerfile, that would obviously require building the sdist outside of Docker.  How would that be added to the current automation?

* If you do get this to work, I think it'd be a better idea to use a wheel instead of an sdist, as a wheel is more "installation-ready."

**Comment by candleindark** [comment]
@yarikoptic @jwodder Thanks for your inputs. In light of your ideas, I think we can approach this problem by using the multi-stage build feature in Docker/Podman. Essentially, using this feature, we can separate the build environment from the runtime environment and include only the runtime environment in the container image for the project.
